api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/AddPatientToQueueAction.java

```diff
@@ -38,72 +36,68 @@ public class AddPatientToQueueAction implements CustomFormSubmissionAction {
+                    }
+                    queue = queues.get(0);
 
-                List<Concept> allowedStatuses = queueServicesWrapper.getAllowedStatuses(queue);
-                if (!allowedStatuses.isEmpty()) {
-                    queueEntry.setStatus(allowedStatuses.get(0));
-                }
+                    queueEntry = new QueueEntry();
+                    queueEntry.setQueue(queue);
+                    queueEntry.setPatient(patient);
 
-                List<Concept> allowedPriorities = queueServicesWrapper.getAllowedPriorities(queue);
-                if (!allowedPriorities.isEmpty()) {
-                    queueEntry.setPriority(allowedPriorities.get(0));
-                }
-                queueEntry.setStartedAt(encounter.getEncounterDatetime());
-                queueEntry.setVisit(encounter.getVisit());
+                    List<Concept> allowedStatuses = queueServicesWrapper.getAllowedStatuses(queue);
+                    if (!allowedStatuses.isEmpty()) {
+                        queueEntry.setStatus(allowedStatuses.get(0));
+                    }
 
-                QueueEntrySearchCriteria searchCriteria = new QueueEntrySearchCriteria();
-                searchCriteria.setPatient(patient);
-                searchCriteria.setQueues(Collections.singletonList(queue));
-                List<QueueEntry> queueEntries = queueServicesWrapper.getQueueEntryService().getQueueEntries(searchCriteria);
-                boolean isPatientAlreadyOnTheQueue = false;
-                for (QueueEntry entry : queueEntries) {
-                    if (QueueUtils.datesOverlap(entry.getStartedAt(), entry.getEndedAt(), queueEntry.getStartedAt(), queueEntry.getEndedAt())) {
-                        isPatientAlreadyOnTheQueue = true;
-                        break;
+                    List<Concept> allowedPriorities = queueServicesWrapper.getAllowedPriorities(queue);
+                    if (!allowedPriorities.isEmpty()) {
+                        queueEntry.setPriority(allowedPriorities.get(0));
                     }
+                    queueEntry.setStartedAt(encounter.getEncounterDatetime());
+                    queueEntry.setVisit(encounter.getVisit());
 
-                }
-                if (!isPatientAlreadyOnTheQueue) {
-                    try {
-                        queueServicesWrapper.getQueueEntryService().saveQueueEntry(queueEntry);
-                    } catch (Exception e) {
-                        throw new RuntimeException("AddPatientToQueueAction failed to add patient to queue", e);
+                    QueueEntrySearchCriteria searchCriteria = new QueueEntrySearchCriteria();
+                    searchCriteria.setPatient(patient);
+                    searchCriteria.setQueues(Collections.singletonList(queue));
+                    searchCriteria.setIsEnded(null);
```

Comment:
Yes, this is currently needed and not great as you point out.  I have a PR in the queue module to address this so this should not be needed after / assuming that is merged.